Fix codecs

voith opened the pull request:
Issue discussed here https://github.com/scrapinghub/frontera/issues/211#issuecomment-251931413
Todo List
- [X] Fix msgpack codec
- [x] Fix json codec
- [x] Integration test with Hbase backend(manually)

This PR fixes #211

Other things done in this besides the todo list:
- Added two methods `_convert` and `reconvert` in json codec. These are needed as JSONEncoder accepts strings only as unicode. Method `convert` converts objects recursively to unicode and saves their type.
- made the requirement of msgpack >=0.4 as only versions greater than 0.4 support the changes made in this PR.
- fixed a buggy test case in `test_message_bus_backend` which got exposed after fixing the codecs.

voith commented:
Build is still failing. need https://github.com/scrapinghub/frontera/pull/235 to fix the build error.

voith commented:
managed to get the build passing by temporarily merging #235. Here is the build https://travis-ci.org/scrapinghub/frontera/builds/183326040
The msg-pack and json codecs have been fixed. What is remaining now is to run an integration test manually on the Hbase Backend.

sibiryakov commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -3,29 +3,59 @@
+from w3lib.util import to_unicode, to_bytes
+
+
+def _convert(obj):
```
this is not entirely about conversion, but encoding of types preservation, I propose to change the name to `_encode_recursively()`

voith replied:
I agree. Will do.

voith replied:
done

sibiryakov commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -3,29 +3,59 @@
+    return 'other', obj
+
+
+def _reconvert(obj):
```
the same here

voith replied:
will do

voith replied:
done

sibiryakov commented on frontera/contrib/backends/remote/codecs/msgpack.py:
```diff
@@ -6,31 +6,11 @@
-        elif hasattr(obj, '__dict__'):
-            return serialize(obj.__dict__)
-        else:
-            return None
```
This function skips silently the object if it's not possible to serialize. I propose to keep this behavior, but make it throw a warning.

voith replied:
I have added this function back except with a minor change. I have excluded the encoding part for `six.text_type` as this will be handled now by msgpack

sibiryakov commented on frontera/contrib/backends/remote/codecs/msgpack.py:
```diff
@@ -42,28 +22,29 @@ def __init__(self, request_model, *a, **kw):
 
     def encode_add_seeds(self, seeds):
-        return packb([b'as', [_prepare_request_message(seed) for seed in seeds]])
+        return packb([b'as', [_prepare_request_message(seed) for seed in seeds]], use_bin_type=True, encoding="utf-8")
```
according to short msgpack-python documentation, it's recommended to use encoding argument for ``unpackb`` not for ``packb``

sibiryakov replied:
https://pypi.python.org/pypi/msgpack-python

sibiryakov commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -45,6 +75,10 @@ def __init__(self, request_model, *a, **kw):
         super(Encoder, self).__init__(request_model, *a, **kw)
 
+    def encode(self, obj):
+        converted = _convert(obj)
```
Currently, only meta requires type information to be stored along with the content. Type information cause message size growth, and introduces cpu overhead on parsing. So, I suggest to reimplement this to make it efficient, the same is (and especially) true for msgpack. I don't care much about json codec performance, but msgpack is very important one to be efficient, because it's used in large deployments.

voith replied:
Won't this error occur for headers, cookies or in fact all dict type objects? Right now issues are only reported for meta.

Also any suggestions to make this efficient for msgpack?

voith replied:
This is something I haven't fixed yet. @Preetwinder Your input here would be of great value

sibiryakov commented on tests/test_codecs.py:
```diff
@@ -20,7 +24,8 @@ def check_request(req1, req2):
     enc = encoder(Request, send_body=True)
     dec = decoder(Request, Response)
-    req = Request(url="http://www.yandex.ru",method=b'GET', meta={b"test": b"shmest"}, headers={b'reqhdr': b'value'})
+    req = Request(url="http://www.yandex.ru", method=b'GET',
```
I suggest to explicitly check the type for string objects in meta, to prove that your update codecs code really respects the type and preserves it.

sibiryakov commented:
also @Preetwinder please have a look. Your opinion would much appreciated.

voith commented:
@Preetwinder Since this PR overwrites some of your changes, your input here would be highly appreciated.

chuckus commented:
@voith Thanks for your hard work! I've reviewed the code and it looks good, like the idea of using tuples for the encoding to reduce the payload size as opposed to a dict (or at least that's what I think you're doing).

voith commented:
@chuckus Thanks for your feedback. My idea to use tuples was to create a generator, But unfortunately, `JSONEncoder`  doesn't support generators, so I had to use lists in some places. So the code doesn't really achieve the desired optimization.

voith commented:
rebased. @sibiryakov Can I get a second review? I have still not implemented your suggestion for optimization. I'm waiting for your reply on the comments that I have left behind in the code review

sibiryakov commented on frontera/contrib/backends/remote/codecs/msgpack.py:
```diff
@@ -6,11 +6,29 @@
+        elif hasattr(obj, '__dict__'):
+            return serialize(obj.__dict__)
+        else:
+            return None
```
let's make it generate warning, that object of some non-serializable type was skipped.

sibiryakov commented on tests/test_codecs.py:
```diff
@@ -5,12 +5,38 @@
 import pytest
 
 
+def _compare_dicts(dict1, dict2):
```
is there any reason to have such complex machinery for comparing dicts?

voith replied:
Just wanted to be a little bit precise

sibiryakov replied:
you should understand that having such deep comparisons would require extra processor cycles, and in Frontera we tend to save them and avoid doing unnecessary computation if possible. so please think again if this is really needed. may be try to find an example.

lopuhin replied:
For me it was not immediately obvious, but I think the main difference from built-in dict comparison is this line https://github.com/scrapinghub/frontera/pull/233/files#diff-252cc4202700183e8f0c782fb28dbfc2R34, so it will catch e.g. python 2 dicts with equal str and unicode values, int/float, etc.

sibiryakov commented on tests/test_codecs.py:
```diff
@@ -5,12 +5,38 @@
+    dicts_are_equal = True
+    for key in dict1.keys():
+        if type(dict1[key]) is dict:
+            dicts_are_equal = dicts_are_equal and _compare_dicts(dict1[key], dict2[key])
```
here you could return False earlier in case of inequality, there is no need to iterate till the end

sibiryakov commented:
otherwise looks good!

Preetwinder commented:
LGTM.

voith commented:
@Preetwinder Thanks for your review. @sibiryakov Is there anything else that needs to be done to get this merged?

Preetwinder commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -101,52 +135,51 @@ def __init__(self, request_model, response_model, *a, **kw):
     def _response_from_object(self, obj):
-        url = to_native_str(obj[b'url'])
+        url = obj['url']
         request = self._request_model(url=url,
```
Are the keys guaranteed to be native strings for this and all the following `obj` accesses?

voith replied:
have a look at this line https://github.com/scrapinghub/frontera/pull/233/files?diff=split#diff-6cd48fdd254557ffc1b50d60a0323364R154. The previous method called in the decoder was `dict_to_bytes` which was changing the native_string types to bytes. But the encoder was never changed and it used native_Strings. Since the encoder takes care that it uses native strings the decoder need not worry about any native_string conversion.

sibiryakov replied:
If url had type 'bytes' then after decoding it will have type bytes, and in Frontera URLs have to be native strings (when in process).

voith replied:
> If url had type 'bytes' then after decoding it will have type bytes

You are right about the fact that this PR will save the type of the object it was originally given to the codec and on on decoding it will return  the original type of value.

> and in Frontera URLs have to be native strings

If this is the case than such type checking and conversion should be done somewhere else, Not In a codec. 

>  so maybe we still need a preventive conversion to native strings

A codec should never change the type of the object it was originally inputed with. Forcing such a mechanism in codec is a hacky way of getting around this problem.

sibiryakov replied:
> A codec should never change the type of the object it was originally inputed with. Forcing such a mechanism in codec is a hacky way of getting around this problem.

Agree, it's not a codec's responsibility. But then we would need to check the rest of the code and make sure we're using native strings for URLs. URLs could appear in crawling strategy (sw), backend or message bus. Because such a change could break the code relying on URLs present as native strings.

sibiryakov replied:
Ok, false alarm probably
https://github.com/scrapinghub/frontera/blob/master/frontera/core/models.py#L28
any thoughts @Preetwinder ?

Preetwinder replied:
Yes, URLs shouldn't be a problem, they are guaranteed to be native strings. @voith I understand. The reason for using dict_to_bytes was that the JSON encoder returns unicode strings and we wanted meta, headers, etc to be bytes. But it seems the new function takes care of that.

Preetwinder commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -3,29 +3,59 @@
+    elif isinstance(obj, (list, tuple)):
+        return type(obj).__name__, [_encode_recursively(item) for item in obj]
+    return 'other', obj
+
```
Isn't storing the types 'other, and 'dict' redundant, since the JSON encoder will take care of these types?

voith replied:
A dict can can have nested dicts or other  kinds of objects. Since we are doing some kind of conversion here, we want to save the type here for successful de-conversion.  Json Encoder taking care is not sufficient. We need to successfully Decode this to the original form for which we need to save type.
Although this implementation currently only converts strings, list and dicts it can be easily extended for something like sets. 
If you have any other implementation in mind, please let me know.

sibiryakov replied:
Still it's not clear why 'dict' and 'other' is needed? Please be more specific. My concern is connected with integration with other components. Imagine if somebody else decide to implement this codec in other language. The less type-related information we store here, the easier it would be.

sibiryakov replied:
Yes, thanks for the explanation, now it's much clearer. Ok, let's leave the dict there. But could you leave the comment in the code explaining why it is, at least in short form.

Preetwinder replied:
Thanks for the explanation, I get it.

voith replied:
@sibiryakov will add a docstring.

voith commented:
@Preetwinder thank you for your review. I will address your concerns tonight

voith commented:
sorry for keeping this PR on hold. Have been a little busy lately

sibiryakov commented:
NP, I had vacations too ;)

kmike commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -3,29 +3,59 @@
+
+def _encode_recursively(obj):
+    """
+    recursively encodes an object to unicode and store its type
```
a nitpick: you can't encode to unicode because unicode is not an encoding (at least in Python), you can only decode to unicode

voith replied:
@kmike Thanks for the review. 
I had previously named this method `convert`: https://github.com/scrapinghub/frontera/pull/233/commits/33347d9a9f0c5a9e732612350b320ca09309baa8#diff-6cd48fdd254557ffc1b50d60a0323364R12. 

Side Note: I think I should start contributing to scrapy. There's a lot I can learn from your reviews :)

voith commented:
I've also tested this PR manually with the Hbase backend and it works fine for me

voith commented:
@sibiryakov Please don't merge this. I have to apply kmikes suggestion

sibiryakov commented:
I think we're close. Let's clarify all the issues and try to merge it.

voith commented:
@sibiryakov you're concerns are real. I try to address these issues when I get some more time. Finding it hard to manage time! But i'll get back to this as soon as possible

voith commented:
@sibiryakov I have tried my best to address your comments as Best as I possibly could. Please let me know if there's still any doubt. A a bit late here now(IST), so can't apply kmikes suggestion. But I hope I have atleast addressed your concerns.

sibiryakov commented:
Have you tried the crawling in distributed backends mode with Json codec @voith ?

voith commented:
@sibiryakov No I haven't. I only checked that its not blowing up anything. But I have tested msg-pack in distributed mode. I will let you know when I finish testing the json codec in distributed mode

voith commented:
I have changed the names of the methods `_encode_recursively` and  `_decode_recursively` as pointed out by @kmike. I have also added a docstring explaining in brief  the purpose of these methods and their design considerations.

But I didn't really get the time to test json-codec with the distributed backend. If anyone in the community could help me test it than it would be great. In any case it'll take some more time to test this thoroughly.

sibiryakov commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -9,33 +9,49 @@
 
 
-def _decode_recursively(obj):
+def _revert_from_saved_type(obj):
```
IMO revert is used in the wrong context here http://dictionary.cambridge.org/dictionary/english/revert-to-sth, convert_from_saved_type would be better.

voith replied:
will do

voith replied:
renamed

sibiryakov commented:
OK. Finall call before I merge it. @voith please rename the revert_* method name.
Anyone here has any ideas what it could break @lopuhin @kmike @Preetwinder ?

lopuhin commented on frontera/contrib/backends/remote/codecs/json.py:
```diff
@@ -3,29 +3,75 @@
+    first element of the nested tuple as the original type information and
+    the second value to be the converted value. It applies the original type
+    recursively on the object to retrieve the original form of the object.
+    """
```
This is super minor, but I would maybe add a check that ``obj`` has exactly two items, e.g. with ``obj_type, obj_value =  obj``, just in case some list is passed instead of an object returned by ``_convert_and_save_type``

voith replied:
@lopuhin thanks for the review. Please see if [this](https://github.com/scrapinghub/frontera/pull/233/commits/b7f5a2dc3a2d62e5f5940e486e756835aeb8d47e) looks good to you

lopuhin replied:
Great, thanks @voith! 👍

voith commented:
@sibiryakov Here's a suspicious [line](https://github.com/scrapinghub/frontera/blob/master/frontera/contrib/scrapy/converters.py#L84) I found out. I haven't really tested it, but I have a feeling that this will break. I think as part of this PR some clean up is also needed. All strings related to scrapy should be native strings.

Also I did try to test json codec with the distributed backend. But I was having some problems with my hbase setup. I have recently shifted to mac and `hbase-docker` doesn't seem to work well with it. Allow me some time to work on this. But I can't promise anything. Or if anyone else has the time than fell free to continue this PR.

sibiryakov commented:
Yeah @voith I noticed that. But that was broken before you, and I don't see how it is connected to this PR.

voith commented:
@sibiryakov You're probably right. I wasn't entirely sure. But give me a day to manually test everything. I managed to setup hbase locally. I will verify and let you know

sibiryakov commented:
Sure @voith

voith commented:
@sibiryakov I have finished testing json codec in distributed mode. I can confirm that everything works. The only error I was getting has been fixed in #252. If i apply that patch to this branch than everything works fine. Also that patch has nothing to do with this PR. That bug existed before I made this PR. This PR should be merged after #252 is merged.

voith commented:
I expect some merge conflicts in this branch if #252 gets merged first. I will resolve them later

sibiryakov commented:
ok, rebase it pls.

voith commented:
@sibiryakov I have rebased the PR and fixed merge conflicts.
I have also squashed the commits into two: one for msg-pack and the other for json

sibiryakov commented:
Congratulations and thank you @voith @Preetwinder and others! 🍺 🍺 🍕

voith commented:
Thanks to everyone who reviewed this PR and helped me make the code better! :)

sibiryakov commented:
thank you for bearing all this!

voith commented:
No problem @sibiryakov. I personally learnt a lot from this PR. 
 In fact if any further issues arise due to this PR, I will fix them myself.
